src/com/facebook/buck/worker/WorkerProcess.java

```diff
@@ -166,11 +221,91 @@ public synchronized void close() {
           e,
           "Error while trying to close the worker process %s.",
           Joiner.on(' ').join(processParams.getCommand()));
+    } finally {
+      watchdogShutdownSignal.countDown();
     }
   }
 
   @VisibleForTesting
-  void setProtocol(WorkerProcessProtocol.CommandSender protocolMock) {
+  void launchForTesting(WorkerProcessProtocol.CommandSender protocolMock) {
     this.protocol = protocolMock;
+    this.readerThread.start();
+    this.watchdogThread.start();
+    handshakePerformed = true;
+  }
+
+  private void processNextCommandResponse() throws Throwable {
+    Preconditions.checkState(
+        protocol != null,
+        "Tried to submit a job to the worker process before the handshake was performed.");
+
+    WorkerProcessProtocol.CommandResponse commandResponse =
+        this.protocol.receiveNextCommandResponse();
+    SettableFuture<Integer> result = commandExitCodes.remove(commandResponse.getCommandId());
```

Comment:
This method is called in thread readerLoop, a different thread from submitJob, and both thread access commandExitCodes. Should we protect it?

Reply:
`commandExitCodes` is a `ConcurrentHashMap` so there's no need. Though honestly the threading here is a bit complex because of the watchdog. I updated the implementation to get rid of the watchdog (it was mostly for paranoia) and add a volatile boolean for the shutdown signal as you suggested. Also added a test for worker death.

The locking is still a bit more complex than I'd prefer, so I'm open to suggestions here - there's a weird internal vs external locking of data - externally `this` is used to lock `submitJob` & `close`, but the reader thread uses `readerThread` to be woken up when there is something to do.

Reply:
I thin that's Ok.